### DISCUSSION TITLE
关于内存使用

### DISCUSSION THREAD

**Discussion by tyycssy007** [kkkgo/PaoPaoDNS#48] [问答和帮助] [answered]
Hi 您好。我使用了两个月，非常棒。不过我有个问，我的机器是群晖的716+II，CPU是四核的N3160，内存加到了8g，但是跑了debug.sh后显示如下：
[info] ========== env info ==========                                            
====ENV TEST====                                                                 
MEM:100k 200k 200 16mb                                                           
prefPC:1                                                                         
CORES:-1-                                                                        
POWCORES:-1-                                                                     
ulimit :-1048576-                                                                
FDLIM :-1-                                                                       
TZ:-Asia/Shanghai-                                                               
UPDATE:-weekly-                                                                  
DNS_SERVERNAME:-PaoPaoDNS,blog.03k.org-                                          
SERVER_IP:-none-                                                                 
ETHIP:-192.168.2.3-                                                              
DNSPORT:-53-                                                                     
SOCKS5:-IP:PORT-                                                                 
CNAUTO:-yes-                                                                     
IPV6:-no-                                                                        
CNFALL:-yes-                                                                     
CUSTOM_FORWARD:-IP:PORT-                                                         
AUTO_FORWARD:-no-                                                                
AUTO_FORWARD_CHECK:-yes-                                                         
USE_MARK_DATA:-no-                                                               
RULES_TTL:-0-                                                                    
CN_TRACKER:-yes-                                                                 
USE_HOSTS:-no-                                                                   
HTTP_FILE:-no-                                                                   
SAFEMODE:--                                                                      
QUERY_TIME:-2000ms-                                                              
ADDINFO:-no-                                                                     
PLATFORM:-Linux 991dba157a8d 3.10.108 #64570 SMP Fri Jun 2 23:36:41 CST 2023 x86_
64 Linux-                                                                        
====ENV TEST====                                                                 
[info] ========== process info ==========                                        
PID   USER     TIME  COMMAND                                                     
    1 root      0:00 {init.sh} /bin/sh /usr/sbin/init.sh                         
   21 root      0:00 crond                                                       
  100 root      0:00 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt.to 
  103 root      0:00 mosdns start -d /tmp -c mosdns.yaml                         
  122 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh             
  123 root      0:00 {data_update.sh} /bin/sh /usr/sbin/data_update.sh           
  132 root      0:00 redis-server unixsocket:/tmp/redis.sock                     
  237 root      0:00 unbound -c /tmp/unbound_forward.conf                        
  239 root      0:00 unbound -c /tmp/unbound_raw.conf                            
  265 root      0:00 inotifywait -e modify,delete /etc/unbound/named.cache /data 
  275 root      0:00 {data_update.sh} /bin/sh /usr/sbin/data_update.sh           
  276 root      0:00 curl -4L --connect-timeout 10 -s https://raw.githubusercont 
  277 root      0:00 grep -Eo [0-9A-Za-z]{64}                                    
  279 root      0:00 head -1                                                     
280 root      0:00 /bin/sh                                                     
  286 root      0:00 {debug.sh} /bin/sh /usr/sbin/debug.sh                       
  291 root      0:00 ps -ef                                                      
[info] ========== cn list info ==========                                        
domain:whoami.ds.akahelp.net                                                     
domain:whoami.03k.org                                                            
[info] ========== reids info ==========                                          
used_memory_human:15.93M                                                         
used_memory_rss_human:23.28M                                                     
used_memory_peak_human:15.93M                                                    
total_system_memory_human:7.72G                                                  
used_memory_lua_human:31.00K                                                     
used_memory_vm_total_human:63.00K                                                
used_memory_scripts_human:181B                                                   
maxmemory_human:16.00M                                                           
(integer) 59830

显示最大使用的内存只有16M，即使重新安装最新版的镜像也是如此，所以导致我跑缓存是能是prefPC:1。请问该如何解决，烦请指点，感激不尽。

**Comment by kkkgo** [comment]
看一下启动日志

**Reply by tyycssy007** [reply]
<html>
<body>


content
--
2023/09/28 21:26:07 | stderr | Watches established.
2023/09/28 21:26:07 | stderr | Setting up watches.
2023/09/28 21:26:06 | stdout | /data/trackerslist.txt MODIFY
2023/09/28 21:26:01 | stdout | /usr/sbin/trackerslist.txt.xz Update OK.
2023/09/28 21:26:01 | stdout | /tmp/trackerslist.txt.xz.download Download OK.
2023/09/28 21:26:01 | stderr | 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0 100  2240  100  2240    0     0   2276      0 --:--:-- --:--:-- --:--:--  2276
2023/09/28 21:26:00 | stderr | Dload  Upload   Total   Spent    Left  Speed
2023/09/28 21:26:00 | stderr | % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
2023/09/28 21:26:00 | stdout | oldsum:b6d86d63abc0be9554427bbf8da97f16243432aab1a158bf6d23c20c33838085
2023/09/28 21:26:00 | stdout | newsum:833fa89f8e30eb57f961fdc70ac2c4f0c440bda1edbb56bc57e388836b867b20
2023/09/28 21:26:00 | stdout | /usr/sbin/trackerslist.txt.xz diff sha256sum, update...
2023/09/28 21:25:59 | stderr | Watches established.
2023/09/28 21:25:59 | stderr | Setting up watches.
2023/09/28 21:25:59 | stdout | 436 root      0:00 mosdns start -d /tmp -c mosdns.yaml
2023/09/28 21:25:58 | stderr | killall: mosdns: no process killed
2023/09/28 21:25:58 | stdout | mosdns reload...
2023/09/28 21:25:58 | stderr | killall: mosdns: no process killed
2023/09/28 21:25:58 | stdout | 103 root      0:00 [mosdns]
2023/09/28 21:25:58 | stdout | 103 root      0:00 mosdns start -d /tmp -c mosdns.yaml
2023/09/28 21:25:57 | stdout | 2023-09-28 21:25:57 CST
2023/09/28 21:25:57 | stdout | /data/Country-only-cn-private.mmdb MODIFY
2023/09/28 21:25:52 | stdout | /usr/sbin/Country-only-cn-private.mmdb.xz Update OK.
2023/09/28 21:25:52 | stdout | /tmp/Country-only-cn-private.mmdb Download OK.
2023/09/28 21:25:52 | stderr | 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0   0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0   0 49084    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0  11 49084   11  5590    0     0   1939      0  0:00:25  0:00:02  0:00:23  1939  37 49084   37 18181    0     0   4541      0  0:00:10  0:00:04  0:00:06  4541  42 49084   42 20976    0     0   4193      0  0:00:11  0:00:05  0:00:06  4193  57 49084   57 27968    0     0   4751      0  0:00:10  0:00:05  0:00:05  6321  71 49084   71 34960    0     0   5165      0  0:00:09  0:00:06  0:00:03  7318  79 49084   79 39157    0     0   5033      0  0:00:09  0:00:07  0:00:02  6854  85 49084   85 41952    0     0   4659      0  0:00:10  0:00:09  0:00:01  4754  88 49084   88 43345    0     0   4333      0  0:00:11  0:00:10  0:00:01  4473  99 49084   99 48935    0     0   4340      0  0:00:11  0:00:11 --:--:--  3892 100 49084  100 49084    0     0   4309      0  0:00:11  0:00:11 --:--:--  3057
2023/09/28 21:25:40 | stderr | Dload  Upload   Total   Spent    Left  Speed
2023/09/28 21:25:40 | stderr | % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
2023/09/28 21:25:40 | stdout | oldsum:8de386e9d4823efb29b9002349517010f4530863e05d5d84acf95139d12610b1
2023/09/28 21:25:40 | stdout | newsum:7f84e80c420a85169ef450f5299dd2b9cfb99d48c72071857268f824e14a0943
2023/09/28 21:25:40 | stdout | /usr/sbin/Country-only-cn-private.mmdb.xz diff sha256sum, update...
2023/09/28 21:16:48 | stdout | 2023-09-28 21:16:48 CST
2023/09/28 21:16:48 | stdout | used_memory_human:15.91M
2023/09/28 21:16:45 | stderr | Watches established.
2023/09/28 21:16:45 | stderr | Setting up watches.
2023/09/28 21:16:45 | stdout | 239 root      0:00 unbound -c /tmp/unbound_raw.conf
2023/09/28 21:16:45 | stdout | 237 root      0:00 unbound -c /tmp/unbound_forward.conf
2023/09/28 21:16:44 | stderr | killall: unbound: no process killed
2023/09/28 21:16:44 | stdout | unbound reload...
2023/09/28 21:16:44 | stdout | 163 root      0:00 unbound -c /tmp/unbound_raw.conf -p
2023/09/28 21:16:44 | stdout | 161 root      0:00 unbound -c /tmp/unbound_forward.conf -p
2023/09/28 21:16:43 | stdout | /etc/unbound/named.cache Update OK.
2023/09/28 21:16:43 | stdout | /etc/unbound/named.cache MODIFY
2023/09/28 21:16:43 | stdout | /tmp/named.cache Download OK.
2023/09/28 21:16:43 | stderr | 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0 100  3315  100  3315    0     0   4339      0 --:--:-- --:--:-- --:--:--  4339
2023/09/28 21:16:43 | stderr | Dload  Upload   Total   Spent    Left  Speed
2023/09/28 21:16:43 | stderr | % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
2023/09/28 21:16:43 | stdout | oldsum:9ff7963fd8b39b01a5cb48bf46af364f
2023/09/28 21:16:43 | stdout | newsum:ac281ab5712d761d1a4e7a7224b89666
2023/09/28 21:16:43 | stdout | /etc/unbound/named.cache diff sha256sum, update...
2023/09/28 21:16:41 | stdout | 2023-09-28 21:16:41 CST
2023/09/28 21:16:39 | stderr | WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/community: could not connect to server (check repositories file)
2023/09/28 21:16:39 | stderr | WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/main: could not connect to server (check repositories file)
2023/09/28 21:14:41 | stderr | WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/community: could not connect to server (check repositories file)
2023/09/28 21:14:41 | stderr | WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/main: Permission denied
2023/09/28 21:11:11 | stderr | Watches established.
2023/09/28 21:11:11 | stderr | Setting up watches.
2023/09/28 21:11:11 | stdout | 132:M 28 Sep 2023 21:11:11.339 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
2023/09/28 21:11:11 | stdout | 132:C 28 Sep 2023 21:11:11.335 # systemd supervision requested or auto-detected, but Redis is compiled without libsystemd support!
2023/09/28 21:11:11 | stdout | 132:C 28 Sep 2023 21:11:11.335 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
2023/09/28 21:11:11 | stdout | 131 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
2023/09/28 21:11:11 | stdout | 130 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
2023/09/28 21:11:11 | stdout | 129 root      0:00 ps
2023/09/28 21:11:11 | stdout | 128 root      0:00 sleep 186
2023/09/28 21:11:11 | stdout | 123 root      0:00 {data_update.sh} /bin/sh /usr/sbin/data_update.sh
2023/09/28 21:11:11 | stdout | 122 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
2023/09/28 21:11:11 | stdout | 120 root      0:00 unbound -c /tmp/unbound_raw.conf -p
2023/09/28 21:11:11 | stdout | 103 root      0:00 mosdns start -d /tmp -c mosdns.yaml
2023/09/28 21:11:11 | stdout | 102 root      0:00 unbound -c /tmp/unbound_forward.conf -p
2023/09/28 21:11:11 | stdout | 100 root      0:00 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt.toml
2023/09/28 21:11:11 | stdout | 21 root      0:00 crond
2023/09/28 21:11:11 | stdout | 1 root      0:00 {init.sh} /bin/sh /usr/sbin/init.sh
2023/09/28 21:11:11 | stdout | PID   USER     TIME  COMMAND
2023/09/28 21:11:11 | stderr | 2023-09-28T21:11:11.317+0800 INFO main config loaded {"file": "mosdns.yaml"}
2023/09/28 21:11:11 | stderr | 2023-09-28T21:11:11.312+0800 INFO working directory changed {"path": "/tmp"}
2023/09/28 21:11:11 | stdout | Apply trackerslist...
2023/09/28 21:11:11 | stdout | ====ENV TEST====
2023/09/28 21:11:11 | stdout | PLATFORM:-Linux 991dba157a8d 3.10.108 #64570 SMP Fri Jun 2 23:36:41 CST 2023 x86_64 Linux-
2023/09/28 21:11:11 | stdout | ADDINFO:-no-
2023/09/28 21:11:11 | stdout | QUERY_TIME:-2000ms-
2023/09/28 21:11:11 | stdout | SAFEMODE:--
2023/09/28 21:11:11 | stdout | HTTP_FILE:-no-
2023/09/28 21:11:11 | stdout | USE_HOSTS:-no-
2023/09/28 21:11:11 | stdout | CN_TRACKER:-yes-
2023/09/28 21:11:11 | stdout | RULES_TTL:-0-
2023/09/28 21:11:11 | stdout | USE_MARK_DATA:-no-
2023/09/28 21:11:11 | stdout | AUTO_FORWARD_CHECK:-yes-
2023/09/28 21:11:11 | stdout | AUTO_FORWARD:-no-
2023/09/28 21:11:11 | stdout | CUSTOM_FORWARD:-IP:PORT-
2023/09/28 21:11:11 | stdout | CNFALL:-yes-
2023/09/28 21:11:11 | stdout | IPV6:-no-
2023/09/28 21:11:11 | stdout | CNAUTO:-yes-
2023/09/28 21:11:11 | stdout | SOCKS5:-IP:PORT-
2023/09/28 21:11:11 | stdout | DNSPORT:-53-
2023/09/28 21:11:11 | stdout | ETHIP:-192.168.2.3-
2023/09/28 21:11:11 | stdout | SERVER_IP:-none-
2023/09/28 21:11:11 | stdout | DNS_SERVERNAME:-PaoPaoDNS,blog.03k.org-
2023/09/28 21:11:11 | stdout | UPDATE:-weekly-
2023/09/28 21:11:11 | stdout | TZ:-Asia/Shanghai-
2023/09/28 21:11:11 | stdout | FDLIM :-1-
2023/09/28 21:11:11 | stdout | ulimit :-1048576-
2023/09/28 21:11:11 | stdout | POWCORES:-1-
2023/09/28 21:11:11 | stdout | CORES:-1-
2023/09/28 21:11:11 | stdout | prefPC:1
2023/09/28 21:11:11 | stdout | MEM:100k 200k 200 16mb
2023/09/28 21:11:11 | stdout | ====ENV TEST====
2023/09/28 21:11:11 | stdout | [Warning] use safemem!
2023/09/28 21:11:11 | stdout | [Warning] LOW MEMORY!
2023/09/28 21:11:11 | stdout | MEMSIZE:0
2023/09/28 21:11:11 | stdout | Swap:          6.6G       29.1M        6.6G
2023/09/28 21:11:11 | stdout | -/+ buffers/cache:       899.5M        6.8G
2023/09/28 21:11:11 | stdout | Mem:           7.7G      899.5M        1.3G           0        5.5G           0
2023/09/28 21:11:11 | stdout | total        used        free      shared  buff/cache   available
2023/09/28 21:11:11 | stdout | Swap:          6792          29        6763
2023/09/28 21:11:11 | stdout | -/+ buffers/cache:          899        7004
2023/09/28 21:11:11 | stdout | Mem:           7903         899        1320           0        5684           0
2023/09/28 21:11:11 | stdout | total        used        free      shared  buff/cache   available
2023/09/28 21:11:11 | stdout | ulimit adbove 1000000.
2023/09/28 21:11:11 | stdout | New version install ! Try clean...
2023/09/28 21:11:11 | stdout | images build time : 2023-09-19 13:43:08 UTC
2023/09/28 21:11:11 | stdout | =====PaoPaoDNS docker start=====


</body>
</html>

**Comment by kkkgo** [comment]
你进入到容器内（sh），执行free -m看看

**Reply by tyycssy007** [reply]
显示如下：
 total        used        free      shared  buff/cache   available  
Mem:           7903         961        1235           0        5708           0  
-/+ buffers/cache:          961        6943                                      
Swap:          6792          29        6763

**Reply by kkkgo** [reply]
可以截图吗，这样看不对齐，截图直接ctrl+v粘贴即可

**Reply by tyycssy007** [reply]
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/9511cfac-c085-4811-85f8-ef1db19f6a6e)

**Reply by kkkgo** [reply]
如果对齐没错的话，available一项是0，表示检测到有0可用的内存

**Reply by tyycssy007** [reply]
但是我机器系统显示如下：
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/b79af9c3-851f-4791-a830-4dd841ee6c92)

**Reply by kkkgo** [reply]
容器启动检测的是available作为可用内存，并根据可用内存使用不同的内存策略。比如：
![1695912043862](https://github.com/kkkgo/PaoPaoDNS/assets/4530897/f761e9fc-1708-43be-b712-b00b9c8bcfd3)

**Reply by kkkgo** [reply]
这个是直接原因，你可以在群晖的ssh终端执行free -h看看available是不是也是0

**Reply by tyycssy007** [reply]
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/305b998b-a6f0-4819-b9db-6c3a0cbbc959)
并不是0，很奇怪

**Reply by kkkgo** [reply]
你的容器管理界面是否有限制内存容量的相关选项？

**Reply by kkkgo** [reply]
在群晖ssh终端执行docker stats -a可以查看容器内存分配情况。

**Reply by tyycssy007** [reply]
有这个选项，但是我并未选用
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/1828c120-174e-4168-a746-e1c6ab9cf52e)

**Reply by tyycssy007** [reply]
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/dfdad383-4dff-4342-848f-d4191464200e)
貌似也是正常的

**Reply by kkkgo** [reply]
你尝试停止所有容器，然后单独启动paopaodns，在paopaodns内执行free -m看看available

**Reply by kkkgo** [reply]
包括停止portainer，你可以使用docker stop id来停止容器，docker start id来启动容器

**Reply by tyycssy007** [reply]
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/70adc22b-1fed-4182-9c17-f5f9ee72fcd3)
单独跑还是可用为0

**Reply by kkkgo** [reply]
在其他容器内执行free -m也是0吗

**Reply by tyycssy007** [reply]
adguard果然也是0，那应该是我的系统的问题了？
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/b8abcf48-db80-4520-9e49-9475d89e29a7)

**Reply by kkkgo** [reply]
不好说，可能检测方法不一样。你可以尝试把每个容器的内存限制值都调整到剩余可用内存范围之内的值试试，或者等我弄个虚拟群晖来测试下。

**Reply by tyycssy007** [reply]
好的，我自己再测试一下，有新发现再和您讨教，谢谢了。

**Reply by kkkgo** [reply]
你在容器内执行一下cat /proc/meminfo 贴出来看看

**Reply by tyycssy007** [reply]
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/6b656fa2-9c8b-4626-bd50-5dfa75404c7a)
![image](https://github.com/kkkgo/PaoPaoDNS/assets/146353004/4648a2c8-842b-4028-9790-2cfaefc206f0)
这是执行结果

**Comment by kkkgo** [comment, marked as answer]
根据你的启动信息，你的宿主Linux内核为3.10.108 ，而MemAvailable似乎是Linux 3.14之后才可用，也许这个就是为0的原因（https://blog.famzah.net/2014/09/24/memavailable-metric-for-linux-kernels-before-3-14-in-procmeminfo/ ）。
下个版本我尝试使用其他方法获取可用值以兼容低内核宿主系统。

**Reply by tyycssy007** [reply]
感谢感谢，期待您的下个版本🙏🏻

**Reply by kkkgo** [reply]
该功能已在新版本实现，详细参考更新日志。https://github.com/kkkgo/PaoPaoDNS/discussions/61
